test/test.go

```diff
@@ -0,0 +1,213 @@
+	for _, testDir := range testDirs {
+		testDirPath := filepath.Join(testsRootDir, testDir.Name())
+
+		// read one level of file set only <root_tests_dir>/<test_dir>/file_to_get
+		testFiles, err := filepath.Glob(filepath.Join(testDirPath, "*"))
+		if err != nil {
+			return nil, err
+		}
+
+		// find step-info in dir
+		var stepInfo *models.TestResultStepInfo
+		for _, file := range testFiles {
+			if strings.HasSuffix(file, "step-info.json") {
+				stepInfoFileContent, err := fileutil.ReadBytesFromFile(file)
+				if err != nil {
+					return nil, err
+				}
+				if err := json.Unmarshal(stepInfoFileContent, &stepInfo); err != nil {
+					return nil, err
+				}
+			}
+		}
+
+		// if no step-info.json then skip
+		if stepInfo == nil {
+			continue
+		}
+
+		// get the handler that can manage test type contained in the dir
+		for _, handler := range converters.List() {
+			handler.SetFiles(testFiles)
+			// skip if couldn't find handler for content type
+			if handler.Detect() {
+				xml, err := handler.XML()
+				if err != nil {
+					return nil, err
+				}
+
+				// so here I will have image paths, xml data, and step info per test dir in a bundle info
+				bundle = append(bundle, ResultAssets{
+					XMLContent: xml,
+					ImagePaths: findImages(filepath.Join(testsRootDir, testDir.Name())),
+					StepInfo:   *stepInfo,
+				})
+			}
+		}
+	}
+	return bundle, nil
+}
+
+func checkResponse(resp *http.Response) error {
+	if resp.StatusCode < 200 || 299 < resp.StatusCode {
+		bodyData, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return fmt.Errorf("unsuccessful response code: %d and failed to read body, error: %s", resp.StatusCode, err)
+		}
+		return fmt.Errorf("unsuccessful response code: %d\nbody:\n%s", resp.StatusCode, bodyData)
+	}
+	return nil
+}
+
+// Upload ...
+func (bundle Bundle) Upload(client *http.Client, endpointURL string) error {
+	// generate UUID->filename and UUID->filepath maps and fill them with the assets need to be uploaded
+	assetUploadRequestMap, assetIDPathMap := map[string]string{}, map[string]string{}
+	for _, testResultAsset := range bundle {
+		for _, path := range testResultAsset.ImagePaths {
+			assetUUID := uuid.New().String()
+			assetUploadRequestMap[assetUUID] = filepath.Base(path)
+			assetIDPathMap[assetUUID] = path
+		}
+	}
+
+	// Get an uploadURL map from the server UUID->URL to upload
+	var assetUploadRequestMapData bytes.Buffer
+	if err := json.NewEncoder(&assetUploadRequestMapData).Encode(assetUploadRequestMap); err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodGet, endpointURL, bytes.NewReader(assetUploadRequestMapData.Bytes()))
+	if err != nil {
+		return err
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if err := resp.Body.Close(); err != nil {
+			fmt.Printf("Failed to close body, error: %s\n", err)
+		}
+	}()
+
+	if err := checkResponse(resp); err != nil {
+		return err
+	}
+
+	var assetUploadURLMap map[string]string
```

Comment:
I think you should use `make(map[string]string)` instead.

"Map types are reference types, like pointers or slices, and so the value of m above is nil; it doesn't point to an initialized map. A nil map behaves like an empty map when reading, but attempts to write to a nil map will cause a runtime panic; don't do that."
https://blog.golang.org/go-maps-in-action

Reply:
This is how you need to use for JSON unmarshal :) It is in the next line